src/tcpip/tcpip.c

```diff
@@ -579,9 +579,11 @@ static void read_conn(struct mg_connection *c, struct pkt *pkt) {
     if (s->ack == ack) {
       MG_VERBOSE(("ignoring duplicate pkt"));
     } else {
-      // TODO(cpq): peer sent us SEQ which we don't expect. Retransmit
-      // rather than close this connection
-      mg_error(c, "SEQ != ACK: %x %x %x", seq, s->ack, ack);
+      uint32_t rem_ip;
+      memcpy(&rem_ip, c->rem.ip, sizeof(uint32_t));
+      MG_DEBUG(("SEQ != ACK: %x %x %x", seq, s->ack, ack));
+      tx_tcp((struct mg_tcpip_if *) c->mgr->priv, s->mac, rem_ip, TH_ACK, c->loc.port,
+           c->rem.port, mg_htonl(s->seq), mg_htonl(s->ack), "", 0);
```

Comment:
I got what you are doing (after carefully following the unit test), you send our latest ACKno no matter what the SEQ is.
This looks safe, please indicate what you do in a comment.

I added a test for the other case, in which our ACK gets lost in transit (instead of the sender's segment), so we get a SEQno which is less than what we expect. We must not deliver that to the application, and we should ACK the highest one we've got. (1002 in the test).
It fails, I get PUSH+ACK and 1

Reply:
When the sender resends the 1001 segment, the receiver knows it is a duplicate packet and it just drops it without sending back a reply. In a normal TCP connection, the sender will not wait for the ACK of 1001, but instead it will keep sending the next packets (1002, 1003...), so when an ACK is received for one of these packets, the sender will know that 1001 has been successfully delivered as well. Resending 1001 will not result in an ACK from the receiver, so it shouldn't be checked a reply, like you did in the test. Also, the reason you are getting PUSH+ACK and 1 is because, for simplicity, I am using the same buffer both for sending and receiving data, and since you are not getting any replies, you are reading the content of the TCP segment you just created before.

Reply:
OK, thanks, I'll have to re-read the docs again. The stream nature... 
Yeah, I saw the PSH and imagined that, but the back to 1 looked weird. We might screw the stack in the future so we should check there is a reply, Please add a check for that or don't share buffers if you can't clean 'em up between usages.